### PR TITLE
[Internal][Executor] Refactor batch result for batch engine

### PR DESCRIPTION
**Pull request by PeiwenGaoMS**
# Description

Refactor batch result for batch engine
``` json
{      
  "status": "completed",      
  "total_lines": 5000,     
  "completed_lines": 4900,     
  "failed_lines": 100,      
  "node_status": {      
    "node1.completed": 5000,      
    "node2.completed": 4999,      
    "node2.failed": 1      
  },      
  "start_time  ": "2023-01-01T00:00:00Z",      
  "end_time": "2023-01-01T01:00:00Z",      
  "metrics": {      // aggregation node’s metrics.
    "metric1": "Value1",      
    "metric2": "Value2"      
  },      
  "system_metrics": {      
    "completion_tokens": 10,      
    "prompt_tokens": 250,      
"total_tokens": 260,
"duration ": 29720.352985,	// By second
  },      
  "error_summary": {      
    "failed_user_error_lines": 50,      
    "failed_system_error_lines": 50,      
    "error_list": [     
      {     
        "line_number": 0,    
        "error": { }  // Detail error dict, contains error code/message/message format/additional info and so on
      },    
      {     
        "line_number": 1,    
        "error": { }  // Detail error dict, contains error code/message/message format/additional info and so on
      }    
    ]      
  }      
}
```

# All Promptflow Contribution checklist:
- [x] **The pull request does not introduce [breaking changes].**
- [ ] **CHANGELOG is updated for new features, bug fixes or other significant changes.**
- [x] **I have read the [contribution guidelines](../CONTRIBUTING.md).**
- [ ] **Create an issue and link to the pull request to get dedicated review from promptflow team. Learn more: [suggested workflow](../CONTRIBUTING.md#suggested-workflow).**

## General Guidelines and Best Practices
- [x] Title of the pull request is clear and informative.
- [x] There are a small number of commits, each of which have an informative message. This means that previously merged commits do not appear in the history of the PR. For more information on cleaning up the commits in your PR, [see this page](https://github.com/Azure/azure-powershell/blob/master/documentation/development-docs/cleaning-up-commits.md).

### Testing Guidelines
- [x] Pull request includes test coverage for the included changes.

### CODE REVIEW

File: src/promptflow/promptflow/batch/_result.py [hunk 1]
```diff
@@ -0,0 +1,144 @@
...
+from promptflow._utils.exception_utils import RootErrorCode
+from promptflow._utils.openai_metrics_calculator import OpenAIMetricsCalculator
+from promptflow.contracts.run_info import Status
+from promptflow.executor._result import AggregationResult, LineResult
+
+
+@dataclass
+class LineError:
+    line_number: int
+    error: Dict[str, Any]
```
Comment (thy09) [inline src/promptflow/promptflow/batch/_result.py:19]: Add more comment to explain what is in the dict.

Also I guess it should a mapping? (immutable)

File: src/promptflow/promptflow/batch/_result.py [hunk 2]
```diff
@@ -0,0 +1,144 @@
...
+@dataclass
+class ErrorSummary:
+    """The summary of errors in a batch run."""
+
+    failed_user_error_lines: int
+    failed_system_error_lines: int
+    error_list: List[LineError]
+
+    @staticmethod
+    def summary(line_results: List[LineResult]):
```
Comment (thy09) [inline src/promptflow/promptflow/batch/_result.py:31]: create?

Comment (PeiwenGaoMS) [reply src/promptflow/promptflow/batch/_result.py]: Updated.

File: src/promptflow/promptflow/batch/_batch_engine.py [hunk 3]
```diff
@@ -126,17 +127,18 @@ def _exec_batch(
...
             for r in line_results
             if r.run_info.status == Status.Completed
         ]
-        return BatchResult(
-            outputs=outputs,
-            metrics=aggr_results.metrics,
-            line_results=line_results,
-            aggr_results=aggr_results,
-        )
+        self._persist_outputs(outputs, output_dir)
```
Comment (huaiyan) [inline src/promptflow/promptflow/batch/_batch_engine.py:137]: A question not related to this PR. How do we persist image in output.

File: src/promptflow/tests/executor/e2etests/test_batch_engine.py [hunk 4]
```diff
@@ -162,50 +155,61 @@ def test_batch_with_metrics(self, dev_connections):
...
+            SAMPLE_FLOW, inputs_mapping, connections=dev_connections, return_output_dir=True
+        )
         nlines = get_batch_inputs_line(SAMPLE_FLOW)
-        assert len(batch_result.outputs) == nlines
-        openai_metrics = batch_result.get_openai_metrics()
-        assert "total_tokens" in openai_metrics
-        assert openai_metrics["total_tokens"] > 0
+        outputs = load_jsonl(output_dir / OUTPUT_FILE_NAME)
+        assert len(outputs) == nlines
+        assert batch_result.system_metrics.total_tokens > 0
```
Comment (huaiyan) [inline src/promptflow/tests/executor/e2etests/test_batch_engine.py:194]: Add assert for `prompt_tokens` and `completion_tokens`?

File: src/promptflow/tests/executor/e2etests/test_batch_engine.py [hunk 5]
```diff
@@ -118,27 +110,28 @@ def test_batch_storage(self, dev_connections):
...
+            flow_folder, inputs_mapping, connections=dev_connections, return_output_dir=True
+        )
+
         assert isinstance(batch_result, BatchResult)
         nlines = get_batch_inputs_line(flow_folder)
-        msg = f"Bulk result only has {len(batch_result.line_results)}/{nlines} outputs"
-        assert len(batch_result.outputs) == nlines, msg
-        for i, output in enumerate(batch_result.outputs):
+        assert batch_result.total_lines == nlines
+        assert batch_result.completed_lines == nlines
```
Comment (huaiyan) [inline src/promptflow/tests/executor/e2etests/test_batch_engine.py:120]: Add assert for `start_time` `end_time` `duration` here?

File: src/promptflow/promptflow/batch/_batch_engine.py (hunk 3 above)
Comment (PeiwenGaoMS) [reply src/promptflow/promptflow/batch/_batch_engine.py:137]: In line execution process pool
image

File: src/promptflow/tests/executor/e2etests/test_batch_engine.py (hunk 4 above)
Comment (PeiwenGaoMS) [reply src/promptflow/tests/executor/e2etests/test_batch_engine.py:194]: Updated.

File: src/promptflow/tests/executor/e2etests/test_batch_engine.py (hunk 5 above)
Comment (PeiwenGaoMS) [reply src/promptflow/tests/executor/e2etests/test_batch_engine.py:120]: Updated.

File: src/promptflow/tests/executor/e2etests/test_activate.py [hunk 6]
```diff
@@ -65,52 +77,78 @@ def test_all_nodes_bypassed(self, dev_connections):
...
+        outputs = load_jsonl(output_dir / OUTPUT_FILE_NAME)
+        for i, output in enumerate(outputs):
             expected_outputs = expected_result[i]["expected_outputs"].copy()
             expected_outputs.update({"line_number": i})
             assert output == expected_outputs
 
         # Validate the flow line results
-        for i, line_result in enumerate(result.line_results):
-            self.assert_activate_flow_run_result(line_result, expected_result[i])
+        flow_runs = {k: v for k, v in sorted(mem_run_storage._flow_runs.items(), key=lambda item: item[1].index)}
```
Comment (zzn2) [inline src/promptflow/tests/executor/e2etests/test_activate.py:120]: Will this be better for understanding?
```suggestion
        flow_runs = {k: v for k, v in sorted(mem_run_storage._flow_runs.items(), key=lambda _, v: v.index)}
```

Comment (PeiwenGaoMS) [reply src/promptflow/tests/executor/e2etests/test_activate.py:120]: Updated.

Comment (PeiwenGaoMS) [reply src/promptflow/tests/executor/e2etests/test_activate.py:120]: But it can't work. I reverted it.

**Review by zhengfeiwang — APPROVED** [review]

**Review by D-W- — APPROVED** [review]

**Review by thy09 — APPROVED** [review]

**Pull request merged**